File: cmd/gen-node-defs/repo_root.go
package main

import (
	"os"
	"path/filepath"
)

func findRepoRoot(dir string) string {
	for {
		if _, err := os.Stat(filepath.Join(dir, "nodes")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return ""
		}
		dir = parent
	}
}

File: Categories/extension/repo-root.ts
import * as fs from "fs";
import * as path from "path";

export function resolveRepoRoot(startDir: string | undefined): string | undefined {
  if (!startDir) return undefined;

  let dir = path.resolve(startDir);
  for (;;) {
    if (fs.existsSync(path.join(dir, ".git"))) return dir;
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }

  return path.resolve(startDir);
}
